request to support pytest configured with other naming conventions other than "test_"

Pytest has a configuration option to change the default so that instead of tests being named "test_" you can change it to "check_" or "should_" or similar. If you do that, approvals stops working. If pytest can find and execute your test I would expect approvals to also still work and evaluate whether the test passed or not.

This is the docs for pytest:

https://docs.pytest.org/en/stable/example/pythoncollection.html#change-naming-conventions

With this pytest.ini:

[pytest]
python_functions = *_check

and this test code:

def instance_check():
    result = str(Hiker().instance_answer())
    verify(result)
 
then you get an error like this:

Could not find test method/function. Possible reasons could be:
    1) approvaltests is not being used inside a test function
    2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported)."""
>           raise FrameNotFound(message)
E           approvaltests.approval_exception.FrameNotFound: Could not find test method/function. Possible reasons could be: 
E           1) approvaltests is not being used inside a test function 
E           2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported).

/usr/local/lib/python3.10/site-packages/approvaltests/core/namer.py:77: FrameNotFound

I have this cyber-dojo instance set up with this problem:

https://cyber-dojo.org/kata/edit/1cMGrz

--- a/approvaltests/namer/stack_frame_namer.py
+++ b/approvaltests/namer/stack_frame_namer.py
@@ -2,10 +2,12 @@ import inspect
 import os
 from inspect import FrameInfo
 from typing import Optional, Dict, List
+import fnmatch
 
 from approvaltests.namer.namer_base import NamerBase
 from approvaltests.approval_exception import FrameNotFound
 from approval_utilities.utilities.stack_frame_utilities import get_class_name_for_frame
+from approvaltests.pytest.pytest_config import PytestConfig
 
 
 class StackFrameNamer(NamerBase):
@@ -40,7 +42,24 @@ class StackFrameNamer(NamerBase):
         raise FrameNotFound(message)
 
     @staticmethod
-    def is_test_method(frame: FrameInfo) -> bool:
+    def is_pytest_test(method_name: str) -> bool:
+        pattern = PytestConfig.test_naming_pattern
+        
+        # taken from pytest/python.py (class PyCollector)
+        if method_name.startswith(pattern):
+            return True
+            # Check that name looks like a glob-string before calling fnmatch
+            # because this is called for every name in each collected module,
+            # and fnmatch is somewhat expensive to call.
+        elif ("*" in pattern or "?" in pattern or "[" in pattern) and fnmatch.fnmatch(
+            method_name, pattern
+        ):
+            return True
+        
+        return False
+
+    @staticmethod
+    def is_unittest_test(frame: FrameInfo) -> bool:
         method_name = frame[3]
         local_attributes = frame[0].f_locals
         is_unittest_test = (
@@ -51,10 +70,15 @@ class StackFrameNamer(NamerBase):
             and method_name != "_callTestMethod"
             and method_name != "run"
         )
+        return is_unittest_test
 
-        is_pytest_test = method_name.startswith("test_")
+    @staticmethod
+    def is_test_method(frame: FrameInfo) -> bool:
+        method_name = frame[3]
 
-        return is_unittest_test or is_pytest_test
+        return StackFrameNamer.is_unittest_test(
+            frame
+        ) or StackFrameNamer.is_pytest_test(method_name)
 
     def get_class_name(self) -> str:
         return self.class_name

--- a/approvaltests/namer/stack_frame_namer.py
+++ b/approvaltests/namer/stack_frame_namer.py
@@ -41,22 +41,25 @@ class StackFrameNamer(NamerBase):
 2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported)."""
         raise FrameNotFound(message)
 
+    # change parameter from method_name to pass in the frame
     @staticmethod
-    def is_pytest_test(method_name: str) -> bool:
+    def is_pytest_test(frame: FrameInfo) -> bool:
+        #  def is_pytest_test(method_name: str) -> bool:
+
         patterns = PytestConfig.test_naming_patterns
-        
+
         # taken from pytest/python.py (class PyCollector)
         for pattern in patterns:
-            if method_name.startswith(pattern):
+            if frame[3].startswith(pattern):
                 return True
                 # Check that name looks like a glob-string before calling fnmatch
                 # because this is called for every name in each collected module,
                 # and fnmatch is somewhat expensive to call.
-            elif ("*" in pattern or "?" in pattern or "[" in pattern) and fnmatch.fnmatch(
-                method_name, pattern
-            ):
+            elif (
+                "*" in pattern or "?" in pattern or "[" in pattern
+            ) and fnmatch.fnmatch(frame[3], pattern):
                 return True
-        
+
         return False
 
     @staticmethod
@@ -75,11 +78,11 @@ class StackFrameNamer(NamerBase):
 
     @staticmethod
     def is_test_method(frame: FrameInfo) -> bool:
-        method_name = frame[3]
+        # method_name = frame[3]
 
         return StackFrameNamer.is_unittest_test(
             frame
-        ) or StackFrameNamer.is_pytest_test(method_name)
+        ) or StackFrameNamer.is_pytest_test(frame)
 
     def get_class_name(self) -> str:
         return self.class_name

--- a/approvaltests/namer/stack_frame_namer.py
+++ b/approvaltests/namer/stack_frame_namer.py
@@ -43,18 +43,19 @@ class StackFrameNamer(NamerBase):
 
     @staticmethod
     def is_pytest_test(frame: FrameInfo) -> bool:
+        method_name = frame[3]
         patterns = PytestConfig.test_naming_patterns
 
         # taken from pytest/python.py (class PyCollector)
         for pattern in patterns:
-            if frame[3].startswith(pattern):
+            if method_name.startswith(pattern):
                 return True
                 # Check that name looks like a glob-string before calling fnmatch
                 # because this is called for every name in each collected module,
                 # and fnmatch is somewhat expensive to call.
             elif (
                 "*" in pattern or "?" in pattern or "[" in pattern
-            ) and fnmatch.fnmatch(frame[3], pattern):
+            ) and fnmatch.fnmatch(method_name, pattern):
                 return True
 
         return False

--- a/approvaltests/namer/stack_frame_namer.py
+++ b/approvaltests/namer/stack_frame_namer.py
@@ -41,11 +41,8 @@ class StackFrameNamer(NamerBase):
 2) your test framework is not supported by ApprovalTests (unittest and pytest are currently supported)."""
         raise FrameNotFound(message)
 
-    # change parameter from method_name to pass in the frame
     @staticmethod
     def is_pytest_test(frame: FrameInfo) -> bool:
-        #  def is_pytest_test(method_name: str) -> bool:
-
         patterns = PytestConfig.test_naming_patterns
 
         # taken from pytest/python.py (class PyCollector)
@@ -78,8 +75,6 @@ class StackFrameNamer(NamerBase):
 
     @staticmethod
     def is_test_method(frame: FrameInfo) -> bool:
-        # method_name = frame[3]
-
         return StackFrameNamer.is_unittest_test(
             frame
         ) or StackFrameNamer.is_pytest_test(frame)

--- a/approvaltests/namer/stack_frame_namer.py
+++ b/approvaltests/namer/stack_frame_namer.py
@@ -43,18 +43,19 @@ class StackFrameNamer(NamerBase):
 
     @staticmethod
     def is_pytest_test(method_name: str) -> bool:
-        pattern = PytestConfig.test_naming_pattern
+        patterns = PytestConfig.test_naming_patterns
         
         # taken from pytest/python.py (class PyCollector)
-        if method_name.startswith(pattern):
-            return True
-            # Check that name looks like a glob-string before calling fnmatch
-            # because this is called for every name in each collected module,
-            # and fnmatch is somewhat expensive to call.
-        elif ("*" in pattern or "?" in pattern or "[" in pattern) and fnmatch.fnmatch(
-            method_name, pattern
-        ):
-            return True
+        for pattern in patterns:
+            if method_name.startswith(pattern):
+                return True
+                # Check that name looks like a glob-string before calling fnmatch
+                # because this is called for every name in each collected module,
+                # and fnmatch is somewhat expensive to call.
+            elif ("*" in pattern or "?" in pattern or "[" in pattern) and fnmatch.fnmatch(
+                method_name, pattern
+            ):
+                return True
         
         return False
 

--- a/approvaltests/pytest/pytest_config.py
+++ b/approvaltests/pytest/pytest_config.py
@@ -1,10 +1,10 @@
 class PytestConfig:
-    test_naming_pattern = "test_*"
+    test_naming_patterns = ["test_*"]
 
     @staticmethod
     def set_config(config):
-        PytestConfig.test_naming_pattern = config.getini("python_functions")
+        PytestConfig.test_naming_patterns = config.getini("python_functions")
 
 
-def pytest_config(config):
+def set_pytest_config(config):
     PytestConfig.set_config(config)

--- a/approvaltests/pytest/pytest_config.py
+++ b/approvaltests/pytest/pytest_config.py
@@ -1,0 +1,10 @@
+class PytestConfig:
+    test_naming_pattern = "test_*"
+
+    @staticmethod
+    def set_config(config):
+        PytestConfig.test_naming_pattern = config.getini("python_functions")
+
+
+def pytest_config(config):
+    PytestConfig.set_config(config)

--- a/approvaltests/pytest/pytest_plugin.py
+++ b/approvaltests/pytest/pytest_plugin.py
@@ -1,4 +1,4 @@
-from approvaltests.pytest import pytest_config
+from pytest_config import pytest_config
 
 def pytest_configure(config):
     pytest_config(config)

--- a/approvaltests/pytest/pytest_plugin.py
+++ b/approvaltests/pytest/pytest_plugin.py
@@ -1,4 +1,4 @@
-from pytest_config import pytest_config
+from .pytest_config import set_pytest_config
 
 def pytest_configure(config):
-    pytest_config(config)
+    set_pytest_config(config)

--- a/approvaltests/pytest/pytest_plugin.py
+++ b/approvaltests/pytest/pytest_plugin.py
@@ -1,4 +1,4 @@
-from approvaltests import pytest_config
+from approvaltests.pytest import pytest_config
 
 def pytest_configure(config):
     pytest_config(config)

--- a/approvaltests/pytest/pytest_plugin.py
+++ b/approvaltests/pytest/pytest_plugin.py
@@ -1,0 +1,4 @@
+from approvaltests import pytest_config
+
+def pytest_configure(config):
+    pytest_config(config)

--- a/setup_utils.py
+++ b/setup_utils.py
@@ -31,6 +31,11 @@ def do_the_setup(package_name, package_description, required, extra_requires):
         python_requires=">=3.8",
         packages=find_packages(include=["approvaltests*"]),
         package_data={"approvaltests": ["reporters/reporters.json"]},
+        entry_points={
+            'pytest11': [
+                'approvaltests_pytest = approvaltests.pytest.pytest_plugin',
+            ],
+        },
         install_requires=required,
         extras_require=extra_requires,
         long_description=(get_parent_directory() / "README.md").read_text(),

